contract/src/lib.rs

```diff
@@ -0,0 +1,421 @@
+                    status: String::from("Win"),
+                    date: message.date,
+                    assets: U128(message.assets.0 * 2),
+                };
+
+                self.add_game(game);
+            } else {
+                let game = Game {
+                    player_id: message.player_id.clone(),
+                    status: String::from("Lose"),
+                    date: message.date,
+                    assets: message.assets,
+                };
+
+                self.add_game(game);
+            }
+
+            return PromiseOrValue::Value(U128(0));
+        }
+    }
+}
+
+#[near_bindgen]
+impl Storage {
+    #[init]
+    pub fn new(version: u8) -> Self {
+        Self {
+            version,
+            games: UnorderedMap::new(StorageKey::Games),
+            finished_games: UnorderedMap::new(StorageKey::FinishedGames),
+            assets: UnorderedMap::new(StorageKey::Assets),
+        }
+    }
+
+    pub fn add_game(&mut self, game: Game) {
+        let id = game.get_id();
+
+        let mut player_games =
+            self.games
+                .get(&game.player_id)
+                .unwrap_or(TreeMap::new(StorageKey::GamesWithKey {
+                    game_id: game.player_id.clone(),
+                }));
+
+        let mut finished_games = self
+            .finished_games
+            .get(&game.player_id)
+            .unwrap_or(TreeMap::new(StorageKey::FinishedGamesWithKey {
+                game_id: game.player_id.clone(),
+            }));
+
+        if player_games.contains_key(&id) || finished_games.contains_key(&id) {
+            env::panic_str("Game already exists")
+        };
+
+        if game.status == String::from("Win") {
+            finished_games.insert(&id, &game);
+            self.finished_games.insert(&game.player_id, &finished_games);
+        }
+
+        player_games.insert(&id, &game);
+
+        self.games.insert(&game.player_id, &player_games);
+    }
+
+    pub fn transfer_tokens_to_winner(&mut self, game: Game) {
+        let player_games =
+            self.games
+                .get(&game.player_id)
+                .unwrap_or(TreeMap::new(StorageKey::GamesWithKey {
+                    game_id: game.player_id.clone(),
+                }));
+
+        let finished_games = self
+            .finished_games
+            .get(&game.player_id)
+            .unwrap_or(TreeMap::new(StorageKey::FinishedGamesWithKey {
+                game_id: game.player_id.clone(),
+            }));
+        let id = game.get_id();
+
+        if player_games.contains_key(&id) && !finished_games.contains_key(&id) {
+            env::panic_str("Game already exists")
+        }
+
+        if !player_games.contains_key(&id) {
+            env::panic_str("Game is invalid")
+        }
+
+        let gas_for_next_callback =
+            env::prepaid_gas() - env::used_gas() - FT_TRANSFER_TGAS - RESERVE_TGAS;
+
+        ft_token::ext(AccountId::new_unchecked(String::from("wusn.testnet")))
```

Comment:
usn contract id hardcoded, it should be in the state and initialized in the constructor